bgpd/bgp_mplsvpn.c

```diff
@@ -2454,3 +2484,174 @@ void vpn_leak_postchange_all(void)
 			bgp);
 	}
 }
+
+/* When a bgp vrf instance is unconfigured, remove its routes
+ * from VPN and this vrf could be importing routes from other
+ * bgp vrf instnaces, unimport them.
+ */
+int bgp_vpn_leak_unimport(struct bgp *from_bgp, struct vty *vty)
+{
+	struct listnode *next;
+	struct bgp *to_bgp;
+	const char *tmp_name;
+	char *vname;
+	struct ecommunity *ecom;
+	struct listnode *node;
+	safi_t safi = SAFI_UNICAST;
+	afi_t afi;
+	bool is_vrf_leak_bind;
+	int debug;
+
+	debug = (BGP_DEBUG(vpn, VPN_LEAK_TO_VRF) |
+		     BGP_DEBUG(vpn, VPN_LEAK_FROM_VRF));
+
+	tmp_name = from_bgp->name ? from_bgp->name : VRF_DEFAULT_NAME;
+
+	/* Default instnace can not be removed if VRF leak is enabled
+	 * across any vrf instance.
+	 */
+	if (from_bgp->inst_type == BGP_INSTANCE_TYPE_DEFAULT) {
+		if (is_bgp_vpn_leak_present()) {
+			vty_out(vty,
+			"%% VRF Leak is present, cannot remove default instance.\n");
+			return -1;
+		}
+		return 0;
+	}
+
+	if (from_bgp->inst_type != BGP_INSTANCE_TYPE_VRF)
+		return 0;
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		/* vrf leak is for IPv4 and IPv6 Unicast only */
+		if (afi != AFI_IP && afi != AFI_IP6)
+			continue;
+
+		for (ALL_LIST_ELEMENTS_RO(bm->bgp, next, to_bgp)) {
+			if (from_bgp == to_bgp)
+				continue;
+
+			/* Unimport and remove source vrf from the
+			 * other vrfs import list.
+			 */
+			is_vrf_leak_bind = false;
+			for (ALL_LIST_ELEMENTS_RO(to_bgp->vpn_policy[afi]
+						  .import_vrf, node,
+						  vname)) {
+				if (strcmp(vname, tmp_name) == 0) {
+					is_vrf_leak_bind = true;
+					break;
+				}
+			}
+			if (!is_vrf_leak_bind)
+				continue;
+
+			if (debug)
+				zlog_debug("%s: unimport routes from %s to_bgp %s afi %s import vrfs count %u",
+					   __func__, from_bgp->name_pretty,
+					   to_bgp->name_pretty, afi2str(afi),
+					   to_bgp->vpn_policy[afi].
+						import_vrf->count);
+
+			vrf_unimport_from_vrf(to_bgp, from_bgp, afi, safi);
+
+			/* readd vrf name as unimport removes import vrf name
+			 * from the destination vrf's import list where the
+			 * `import vrf` configuration still exist.
+			 */
+			vname = XSTRDUP(MTYPE_TMP, tmp_name);
+			listnode_add(to_bgp->vpn_policy[afi].import_vrf,
+				     vname);
+			SET_FLAG(to_bgp->af_flags[afi][safi],
+				 BGP_CONFIG_VRF_TO_VRF_IMPORT);
+
+			/* If to_bgp exports its routes to the bgp vrf
+			 * which is being deleted, un-import the
+			 * to_bgp routes from VPN.
+			 */
+			is_vrf_leak_bind = false;
+			for (ALL_LIST_ELEMENTS_RO(to_bgp->vpn_policy[afi]
+						  .export_vrf, node,
+						  vname)) {
+				if (strcmp(vname, tmp_name) == 0) {
+					is_vrf_leak_bind = true;
+					break;
+				}
+			}
+			if (is_vrf_leak_bind)
```

Comment:
Why can't we call vrf_unimport_from_vrf right away and "break" when if (strcmp(vname, tmp_name) == 0) condition is satisfied. Do we really need this boolean variable and associated logic ?

Reply:
Earlier there were more tasks used to perform but have consolidated under vrf_unimport_from_vrf(). I have removed the boolean check and directly called the vrf_unimport.
This is the case when a vrf instance is also exporting its routes to the source vrf instance which is being deleted, unimport them so source vrf export count can be adjusted.